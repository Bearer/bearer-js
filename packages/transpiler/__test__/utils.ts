import * as ts from 'typescript'
export default (sourceFile: ts.SourceFile) => {
  let names: string[] = []
  traverseNode(sourceFile, names)
  function traverseNode(node: ts.Node, names: string[]) {
    if (node.kind === ts.SyntaxKind.Decorator) names.push(node.expression.expression.escapedText)
    ts.forEachChild(node, node => {
      traverseNode(node, names)
    })
  }
  return names
}
