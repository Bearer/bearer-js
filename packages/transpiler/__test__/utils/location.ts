import * as path from 'path'

export const FIXTURES = path.join(__dirname, '..', '__fixtures__')
export const BUILD = path.join(__dirname, '..', '.build/src')

export function UnitFixtureDirectory(dir: string) {
  return path.join(FIXTURES, 'unit', dir)
}

export function BuildUnitFixtureDirectory(dir: string) {
  return path.join(BUILD, 'unit', dir)
}
