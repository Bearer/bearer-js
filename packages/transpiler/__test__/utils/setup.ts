import * as fs from 'fs'
import * as globby from 'globby'
import * as path from 'path'

const buildFolder = path.join(__dirname, '..', '.build/src')

process.env.BEARER_SCENARIO_ID = 'SPONGE_BOB'

globby.sync(['**/*.tsx', '**/*.ts', '**/*.json'], { cwd: buildFolder }).forEach(file => {
  const filePath = path.join(buildFolder, file)
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath)
  }
})
