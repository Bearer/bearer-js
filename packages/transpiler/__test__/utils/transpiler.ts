import Transpiler, { TranpilerOptions } from '../../src/index'

export function TranspilerFactory(options: Partial<TranpilerOptions>): Transpiler {
  const defaults: TranpilerOptions = {
    watchFiles: false,
    buildFolder: '../../../.build/'
  }

  return new Transpiler({
    ...defaults,
    ...options
  })
}
