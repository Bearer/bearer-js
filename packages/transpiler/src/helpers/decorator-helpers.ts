import * as ts from 'typescript'

export function hasPropDecoratedWithName(classNode: ts.ClassDeclaration, decoratorName: string): boolean {
  return Boolean(propDecoratedWithName(classNode, decoratorName).length)
}

export function propDecoratedWithName(
  node: ts.ClassDeclaration,
  decoratorName: string
): Array<ts.PropertyDeclaration | null> {
  const props: Array<ts.PropertyDeclaration | null> = []
  ts.forEachChild(node, node => {
    if (ts.isPropertyDeclaration(node) && hasDecoratorNamed(node as ts.PropertyDeclaration, decoratorName)) {
      props.push(node)
    }
  })
  return props
}

export function hasDecoratorNamed(
  decoratedNode: ts.PropertyDeclaration | ts.ClassDeclaration | ts.MethodDeclaration,
  name: string
): boolean {
  return ts.forEachChild(decoratedNode, anode => {
    return ts.isDecorator(anode) && decoratorNamed(anode, name)
  })
}

export function decoratorNamed(tsDecorator: ts.Decorator, name: string): boolean {
  return ts.forEachChild(tsDecorator, node => {
    return ts.isCallExpression(node) && (node.expression as ts.Identifier).escapedText === name
  })
}

export function getDecoratorProperties(tsDecorator: ts.Decorator, index = 0): ts.ObjectLiteralExpression {
  if (!ts.isCallExpression(tsDecorator.expression)) {
    return ts.createObjectLiteral([])
  }
  const expression = tsDecorator.expression
  const argument = expression.arguments[index]

  if (!ts.isObjectLiteralExpression(argument)) {
    return ts.createObjectLiteral([])
  }
  return ts.createObjectLiteral(argument.properties)
}

export function getExpressionFromLiteralObject<T extends ts.Expression>(
  tsLiteral: ts.ObjectLiteralExpression,
  key: string
): T {
  const found: ts.PropertyAssignment = tsLiteral.properties.find(property => {
    if (ts.isStringLiteral(property.name)) {
      return property.name.text === key
    } else {
      return (property.name as ts.Identifier).escapedText.toString() === key
    }
  }) as ts.PropertyAssignment
  if (found) {
    return found.initializer as T
  }
  return null
}

export function getExpressionFromDecorator<T extends ts.Expression>(
  tsDecorator: ts.Decorator,
  key: string,
  index = 0
): T {
  return getExpressionFromLiteralObject(getDecoratorProperties(tsDecorator, index), key)
}

export function getDecoratorNamed(
  tsDecoratedNode: { decorators?: ts.NodeArray<ts.Decorator> },
  name: string
): ts.Decorator | undefined {
  if (tsDecoratedNode.decorators) {
    return tsDecoratedNode.decorators.find(decorator => decoratorNamed(decorator, name))
  }
}

// tslint:disable-next-line:prefer-array-literal
export function extractStringOptions<T>(objectLitteral: ts.ObjectLiteralExpression, keys: Array<keyof T>): Partial<T> {
  const values: Partial<T> = {}
  keys.map(value => {
    const optionValue = getExpressionFromLiteralObject<ts.StringLiteral>(objectLitteral, value as string)
    if (optionValue) {
      values[value as string] = optionValue && optionValue.text
    }
  })
  return values
}

export function extractArrayOptions<T>(
  objectLitteral: ts.ObjectLiteralExpression,
  keys: (keyof T)[]
): { [P in keyof T]?: string[] } {
  const values = {}
  keys.map(value => {
    const optionValue = getExpressionFromLiteralObject<ts.StringLiteral>(objectLitteral, value as string)
    if (optionValue && ts.isArrayLiteralExpression(optionValue)) {
      values[value as string] = optionValue.elements.map(e => {
        if (ts.isStringLiteral(e)) {
          return e.text
        }
        return []
      })
    }
  })
  return values
}

export function extractBooleanOptions<T>(objectLitteral: ts.ObjectLiteralExpression, keys: (keyof T)[]): Partial<T> {
  const values: Partial<T> = {}
  keys.map(value => {
    const optionValue = getExpressionFromLiteralObject<ts.BooleanLiteral>(objectLitteral, value as string)
    if (optionValue) {
      values[value as string] = optionValue && optionValue.kind === ts.SyntaxKind.TrueKeyword
    }
  })

  return values
}
