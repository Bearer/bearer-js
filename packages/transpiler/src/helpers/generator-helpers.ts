import * as ts from 'typescript'

import { createOrUpdateComponentDidLoad } from '../../src/transformers/bearer'

import { Decorators } from './../constants'
import { CreateFetcherMeta, TAddAutoLoad, TCreateLoadDataCall, TCreateLoadResourceMethod } from './../types'
import { idName } from './name-helpers'

export function createFetcher(meta: CreateFetcherMeta) {
  return ts.createProperty(
    [
      ts.createDecorator(
        ts.createCall(ts.createIdentifier(Decorators.Intent), undefined, [ts.createLiteral(meta.intentName)])
      )
    ],
    undefined,
    meta.intentMethodName,
    undefined,
    undefined,
    undefined
  )
}

function propertyReferenceIdNames(meta: TCreateLoadResourceMethod, metaCollection: TCreateLoadResourceMethod[]) {
  return meta.intentArguments.map(name => {
    const metaInfo = metaCollection.find(meta => meta.propDeclarationName === name)
    if (metaInfo) {
      return ts.createPropertyAssignment(
        idName(name),
        ts.createPropertyAccess(ts.createThis(), metaInfo.propertyReferenceIdName)
      )
    }
    return ts.createPropertyAssignment(name, ts.createPropertyAccess(ts.createThis(), name))
  })
}

export function createLoadResourceMethod(meta: TCreateLoadResourceMethod, metaCollection: TCreateLoadResourceMethod[]) {
  const intentCall = ts.createCall(ts.createPropertyAccess(ts.createThis(), meta.intentMethodName), undefined, [
    ts.createObjectLiteral([
      ts.createPropertyAssignment(
        meta.intentReferenceIdKeyName,
        ts.createPropertyAccess(ts.createThis(), meta.propertyReferenceIdName)
      ),
      ...propertyReferenceIdNames(meta, metaCollection)
    ])
  ])
  const udapteState = ts.createArrowFunction(
    undefined,
    undefined,
    [
      ts.createParameter(
        undefined,
        undefined,
        undefined,
        ts.createObjectBindingPattern([ts.createBindingElement(undefined, undefined, 'data')]),
        undefined,
        ts.createTypeLiteralNode([
          ts.createPropertySignature(undefined, 'data', undefined, meta.typeIdentifier, undefined)
        ]),
        undefined
      )
    ],
    undefined,
    undefined,
    ts.createBlock(
      [
        ts.createIf(
          ts.createIdentifier('data'),
          ts.createBlock([
            ts.createStatement(
              ts.createBinary(
                ts.createPropertyAccess(ts.createThis(), meta.propDeclarationName),
                ts.SyntaxKind.EqualsToken,
                ts.createIdentifier('data')
              )
            )
          ])
        )
      ],
      true
    )
  )
  const promiseHandler = ts.createCall(ts.createPropertyAccess(intentCall, 'then'), undefined, [udapteState])
  return ts.createProperty(
    undefined,
    undefined,
    meta.loadMethodName,
    undefined,
    undefined,
    ts.createArrowFunction(
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      ts.createBlock([ts.createStatement(promiseHandler)], true)
    )
  )
}

export function createLoadDataCall(meta: TCreateLoadDataCall) {
  return ts.createStatement(
    ts.createCall(ts.createPropertyAccess(ts.createThis(), meta.loadMethodName), undefined, undefined)
  )
}

export function addAutoLoad(tsClass: ts.ClassDeclaration, meta: TAddAutoLoad): ts.ClassDeclaration {
  if (meta.autoLoad) {
    return createOrUpdateComponentDidLoad(tsClass, block =>
      ts.updateBlock(block, [
        ...block.statements,
        ts.createIf(
          ts.createPropertyAccess(ts.createThis(), meta.propertyReferenceIdName),
          ts.createBlock([createLoadDataCall(meta)])
        )
      ])
    )
  }
  return tsClass
}
