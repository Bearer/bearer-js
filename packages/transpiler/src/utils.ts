import * as ts from 'typescript'

export function getSourceCode(node): string {
  const resultFile = ts.createSourceFile(
    'tmp.ts',
    '',
    ts.ScriptTarget.Latest,
    /*setParentNodes*/ false,
    ts.ScriptKind.TS
  )
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed
  })

  return printer.printNode(ts.EmitHint.Unspecified, node, resultFile)
}
