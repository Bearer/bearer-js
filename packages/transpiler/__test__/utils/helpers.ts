import * as fs from 'fs'
import * as globby from 'globby'
import * as path from 'path'
import * as ts from 'typescript'

import { TranpilerOptions } from '../../src/index'
import { UnitFixtureDirectory } from '../utils/location'

import { TranspilerFactory } from './transpiler'

const ROOT_DIRECTORY = path.join(__dirname, '..')
const buildSrcFolder = path.join(ROOT_DIRECTORY, '.build/src')

export function cleanBuildFolder() {
  const buildFolder = path.join(__dirname, '..', '.build/src')

  globby.sync(['**/*.tsx', '**/*.ts', '**/*.json'], { cwd: buildFolder }).forEach(file => {
    const filePath = path.join(buildFolder, file)
    if (fs.existsSync(filePath)) {
      fs.unlinkSync(filePath)
    }
  })
}

export function runTransformers(code: string, transformers: ts.TransformerFactory<ts.SourceFile>[]) {
  const sourceFile = ts.createSourceFile('tmp.ts', code, ts.ScriptTarget.Latest)
  const transformed = ts.transform(sourceFile, transformers)

  const printer = ts.createPrinter(
    {
      newLine: ts.NewLineKind.LineFeed
    },
    {
      onEmitNode: transformed.emitNodeWithNotification,
      substituteNode: transformed.substituteNode
    }
  )
  const result = printer.printBundle(ts.createBundle(transformed.transformed))
  transformed.dispose()

  return result
}

export function runUnitOn(name: string, transpilerOptions: Partial<TranpilerOptions> = {}) {
  const srcDirectory = UnitFixtureDirectory(name)

  beforeAll(() => {
    cleanBuildFolder()
    runTranspiler(`__fixtures__/unit/${name}`, transpilerOptions)
  })

  fs.readdirSync(srcDirectory).forEach(file => {
    describe(file.replace(/\.tsx?/, '').replace(/\-/g, ' '), () => {
      it('match snapshot', async done => {
        expect.assertions(1)
        const filePath = path.join(buildSrcFolder, file)
        fs.readFile(filePath, 'utf8', (_e, postContent) => {
          expect({
            postContent,
            file
          }).toMatchSnapshot()
          done()
        })
      })
    })
  })
}

export function runTranspiler(srcFolder: string, transpilerOptions: Partial<TranpilerOptions> = {}) {
  const transpiler = TranspilerFactory({
    buildFolder: '.build',
    srcFolder,
    ROOT_DIRECTORY,
    ...transpilerOptions
  })

  transpiler.run()
}
