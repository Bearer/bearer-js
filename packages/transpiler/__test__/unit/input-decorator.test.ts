import { runTransformers } from '../utils/helpers'
import { UnitFixtureDirectory } from '../utils/location'

import Metadata from '../../src/metadata'
import GatherMetadata from '../../src/transformers/gather-metadata'
import InputDecorator from '../../src/transformers/input-decorator'
import * as fs from 'fs'
import * as path from 'path'
const TEST_NAME = 'input-decorator'

const metadata = new Metadata("bearer", "test")

describe('Input decorator units', () => {
  it("generates stuff", (done) => {
    const srcDirectory = UnitFixtureDirectory(TEST_NAME)

    fs.readdirSync(srcDirectory).forEach(file => {
      const filePath = path.join(srcDirectory, file)

      fs.readFile(filePath, 'utf8', (_e, postContent) => {
        const result = runTransformers(postContent, [GatherMetadata({ metadata }), InputDecorator({ metadata })])
        expect(result).toMatchSnapshot()
        done()
      })
    })
  })
})
