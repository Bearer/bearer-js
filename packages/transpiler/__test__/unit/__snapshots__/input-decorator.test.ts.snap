// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input decorator units complete options match snapshot 1`] = `
Object {
  "file": "complete-options.tsx",
  "postContent": "import { BearerRef, State, Component, Listen, Intent, Watch, Prop } from '@bearer/core';
type Farmer = {
    id: string;
    name: string;
};
type Sponge = {
    id: unknown;
    name: unknown;
};
@Component({
    tag: \\"no-options-action\\",
    styleUrl: \\"noOptions.css\\",
    shadow: true
})
class NoOptionsComponent {
    // won't be injected by the transfomer
    @State()
    refNotOverrided: string;
    @State()
    farmer: BearerRef<Farmer>;
    @Listen(\\"body:bearer-SPONGE_BOB-no-options-farmerSaved\\")
    farmerRefIdChanged(event) {
        if (this.farmerRefId !== event.detail.referenceId) {
            this.farmerRefId = event.detail.referenceId;
        }
        else {
            this._loadFarmer();
        }
    }
    _loadFarmer = () => {
        this.fetcherRetrieveFarmer({ referenceId: this.farmerRefId }).then(({ data }: {
            data: BearerRef<Farmer>;
        }) => {
            this.farmer = data;
        });
    };
    private fetcherRetrieveFarmer;
    @Watch(\\"farmerRefId\\")
    _watchFarmer(newValueName: string) {
        if (newValueName) {
            this._loadFarmer();
        }
    }
    @Prop({ \\"mutable\\": true })
    farmerRefId: string;
    @State()
    aString: BearerRef<string> = 'ok';
    @Listen(\\"body:bearer-SPONGE_BOB-no-options-aStringSaved\\")
    aStringRefIdChanged(event) {
        if (this.aStringRefId !== event.detail.referenceId) {
            this.aStringRefId = event.detail.referenceId;
        }
        else {
            this._loadAString();
        }
    }
    _loadAString = () => {
        this.fetcherRetrieveAString({ referenceId: this.aStringRefId }).then(({ data }: {
            data: BearerRef<string>;
        }) => {
            this.aString = data;
        });
    };
    private fetcherRetrieveAString;
    @Watch(\\"aStringRefId\\")
    _watchAString(newValueName: string) {
        if (newValueName) {
            this._loadAString();
        }
    }
    @Prop({ \\"mutable\\": true })
    aStringRefId: string;
    @State()
    aStringWithoutInitializer: BearerRef<string>;
    @Listen(\\"body:bearer-SPONGE_BOB-no-options-aStringWithoutInitializerSaved\\")
    aStringWithoutInitializerRefIdChanged(event) {
        if (this.aStringWithoutInitializerRefId !== event.detail.referenceId) {
            this.aStringWithoutInitializerRefId = event.detail.referenceId;
        }
        else {
            this._loadAStringWithoutInitializer();
        }
    }
    _loadAStringWithoutInitializer = () => {
        this.fetcherRetrieveAStringWithoutInitializer({ referenceId: this.aStringWithoutInitializerRefId }).then(({ data }: {
            data: BearerRef<string>;
        }) => {
            this.aStringWithoutInitializer = data;
        });
    };
    private fetcherRetrieveAStringWithoutInitializer;
    @Watch(\\"aStringWithoutInitializerRefId\\")
    _watchAStringWithoutInitializer(newValueName: string) {
        if (newValueName) {
            this._loadAStringWithoutInitializer();
        }
    }
    @Prop({ \\"mutable\\": true })
    aStringWithoutInitializerRefId: string;
    @State()
    object: BearerRef<{
        title: string;
    }> = { title: 'Guest' };
    @Listen(\\"body:bearer-SPONGE_BOB-no-options-objectSaved\\")
    objectRefIdChanged(event) {
        if (this.objectRefId !== event.detail.referenceId) {
            this.objectRefId = event.detail.referenceId;
        }
        else {
            this._loadObject();
        }
    }
    _loadObject = () => {
        this.fetcherRetrieveObject({ referenceId: this.objectRefId }).then(({ data }: {
            data: BearerRef<{
                title: string;
            }>;
        }) => {
            this.object = data;
        });
    };
    private fetcherRetrieveObject;
    @Watch(\\"objectRefId\\")
    _watchObject(newValueName: string) {
        if (newValueName) {
            this._loadObject();
        }
    }
    @Prop({ \\"mutable\\": true })
    objectRefId: string;
    @State()
    objectWithoutInitializer: BearerRef<{
        title: string;
    }>;
    @Listen(\\"body:bearer-SPONGE_BOB-no-options-objectWithoutInitializerSaved\\")
    objectWithoutInitializerRefIdChanged(event) {
        if (this.objectWithoutInitializerRefId !== event.detail.referenceId) {
            this.objectWithoutInitializerRefId = event.detail.referenceId;
        }
        else {
            this._loadObjectWithoutInitializer();
        }
    }
    _loadObjectWithoutInitializer = () => {
        this.fetcherRetrieveObjectWithoutInitializer({ referenceId: this.objectWithoutInitializerRefId }).then(({ data }: {
            data: BearerRef<{
                title: string;
            }>;
        }) => {
            this.objectWithoutInitializer = data;
        });
    };
    private fetcherRetrieveObjectWithoutInitializer;
    @Watch(\\"objectWithoutInitializerRefId\\")
    _watchObjectWithoutInitializer(newValueName: string) {
        if (newValueName) {
            this._loadObjectWithoutInitializer();
        }
    }
    @Prop({ \\"mutable\\": true })
    objectWithoutInitializerRefId: string;
    @State()
    spongeBob: BearerRef<Sponge>;
    @Listen(\\"body:bearer-SPONGE_BOB-other-group-patrickWasKilled\\")
    patrickChanged(event) {
        if (this.patrick !== event.detail.referenceId) {
            this.patrick = event.detail.referenceId;
        }
        else {
            this._loadSpongeBob();
        }
    }
    _loadSpongeBob = () => {
        this.fetcherRetrieveSpongeBob({ referenceId: this.patrick }).then(({ data }: {
            data: BearerRef<Sponge>;
        }) => {
            this.spongeBob = data;
        });
    };
    private fetcherRetrieveSpongeBob;
    @Watch(\\"patrick\\")
    _watchSpongeBob(newValueName: string) {
        if (newValueName) {
            this._loadSpongeBob();
        }
    }
    @Prop({ \\"mutable\\": true })
    patrick: string;
    @State()
    spongeBobOverrided: BearerRef<Sponge>;
    @Listen(\\"body:bearer-SPONGE_BOB-no-options-spongeBobOverridedSaved\\")
    refNotOverridedChanged(event) {
        if (this.refNotOverrided !== event.detail.referenceId) {
            this.refNotOverrided = event.detail.referenceId;
        }
        else {
            this._loadSpongeBobOverrided();
        }
    }
    _loadSpongeBobOverrided = () => {
        this.fetcherRetrieveSpongeBobOverrided({ referenceId: this.refNotOverrided }).then(({ data }: {
            data: BearerRef<Sponge>;
        }) => {
            this.spongeBobOverrided = data;
        });
    };
    private fetcherRetrieveSpongeBobOverrided;
    @Watch(\\"refNotOverrided\\")
    _watchSpongeBobOverrided(newValueName: string) {
        if (newValueName) {
            this._loadSpongeBobOverrided();
        }
    }
    constructor() {
        Intent(\\"retrieveFarmer\\")(this, \\"fetcherRetrieveFarmer\\");
        Intent(\\"retrieveAString\\")(this, \\"fetcherRetrieveAString\\");
        Intent(\\"retrieveAStringWithoutInitializer\\")(this, \\"fetcherRetrieveAStringWithoutInitializer\\");
        Intent(\\"retrieveObject\\")(this, \\"fetcherRetrieveObject\\");
        Intent(\\"retrieveObjectWithoutInitializer\\")(this, \\"fetcherRetrieveObjectWithoutInitializer\\");
        Intent(\\"killPatrick\\")(this, \\"fetcherRetrieveSpongeBob\\");
        Intent(\\"retrieveSpongeBobOverrided\\")(this, \\"fetcherRetrieveSpongeBobOverrided\\");
    }
    get SCENARIO_ID() { return \\"SPONGE_BOB\\"; }
    @Prop({ \\"context\\": \\"bearer\\" })
    bearerContext: any;
    @Prop()
    setupId: string;
    componentDidLoad() {
        if (this.setupId) {
            this.bearerContext.setupId = this.setupId;
        }
        if (this.farmerRefId) {
            this._loadFarmer();
        }
        if (this.aStringRefId) {
            this._loadAString();
        }
        if (this.aStringWithoutInitializerRefId) {
            this._loadAStringWithoutInitializer();
        }
        if (this.objectRefId) {
            this._loadObject();
        }
        if (this.objectWithoutInitializerRefId) {
            this._loadObjectWithoutInitializer();
        }
        if (this.refNotOverrided) {
            this._loadSpongeBobOverrided();
        }
    }
}
",
}
`;
